### PR TITLE
Bindu, Supriya | MOBN 1399 | Create a Source under an organization

### PR DESCRIPTION
**Pull request by supriyatw**
# JIRA TICKET NAME:
[JIRA ticket name](https://issues.openmrs.org/browse/<ticket-id>)

# Summary:

### CODE REVIEW

File: src/apps/sources/__test__/components/SourceForm.test.tsx [hunk 1]
```diff
@@ -63,9 +60,25 @@ describe('SourceForm', () => {
...
-       // @ts-ignore
-       expect(getByLabelText('Website').value).toBe('website');
-       // @ts-ignore
-        expect(getByLabelText('Source type').value).toBe('MSF');
    });
 });
+
+describe('Create SourceForm', () => {
+
+    it('should show Source name is required if we dont enter', async () => {
```
Comment (rakesh-1010) [inline src/apps/sources/__test__/components/SourceForm.test.tsx]: @binduak @supriyatw it statement and assertions don't match. Can we modify either of them

File: src/apps/sources/__test__/components/SourceForm.test.tsx [hunk 2]
```diff
@@ -63,9 +60,25 @@ describe('SourceForm', () => {
...
+        let shortCode: HTMLInputElement = getByLabelText(/Short Code/) as HTMLInputElement;
+        let sourceName: HTMLInputElement = getByLabelText(/Source Name/) as HTMLInputElement;
+        let submitButton: HTMLInputElement = getByText('Submit') as HTMLInputElement;
+        await act(async () => {
+            fireEvent.change(shortCode, {target: {value: '10'}});
+            fireEvent.change(sourceName, {target: {value: 'Tustin'}});
+            fireEvent.click(submitButton);
+        });
+        expect(shortCode.value).toBe('10');
+        expect(sourceName.value).toBe('Tustin');
```
Comment (rakesh-1010) [inline src/apps/sources/__test__/components/SourceForm.test.tsx:91]: @supriyatw @binduak  can we avoid using multiple assertions wherever possible. Tests should have one reason to fail, to paint an accurate picture which part is failing.

We can extract the common code in a `before.each` block and have 2 it blocks

File: src/apps/sources/__test__/pages/CreateSourcePage.test.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    it('createSourcePage snapshot test', () => {
+        const {container} = renderUI();
+        expect(container).toMatchSnapshot();
+    });
+    it('should list down all the props of the state', () => {
+        expect(mapStateToProps(state).loading).not.toBeNull();
+        expect(mapStateToProps(state).profile).not.toBeNull();
+        expect(mapStateToProps(state).usersOrgs).not.toBeNull();
+        expect(mapStateToProps(state).newSource).not.toBeNull();
+        expect(mapStateToProps(state).errors).not.toBeNull();
```
Comment (rakesh-1010) [inline src/apps/sources/__test__/pages/CreateSourcePage.test.tsx:63]: can we avoid using multiple assertions wherever possible. Tests should have one reason to fail, to paint an accurate picture which part is failing.

File: src/apps/sources/__test__/pages/CreateSourcePage.test.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import * as React from "react";
+import CreateSourcePage, {mapActionsToProps, mapStateToProps} from "../../pages/CreateSourcePage";
+import {APIOrg, APIProfile} from "../../../authentication";
+import {APISource, Source} from "../../types";
+import {createSourceAction, createSourceDispatchAction} from "../../redux";
+import {act, fireEvent, render} from "@testing-library/react";
```
Comment (rakesh-1010) [inline src/apps/sources/__test__/pages/CreateSourcePage.test.tsx]: can we import this from `src/test-utils.tsx` hoc. That way we don't have import `store`, `provide`, `router`

File: src/apps/sources/__test__/pages/CreateSourcePage.test.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    profile : apiProfile,
+    usersOrgs : [apiOrg],
+    createSourceAction: function createSourceAction() {
+    },
+    loading: true,
+    newSource : source
+};
+
+function renderUI(props: Partial<createSourcePageProps> = {}) {
+    return render(<Provider store={store}>
```
Comment (rakesh-1010) [inline src/apps/sources/__test__/pages/CreateSourcePage.test.tsx]: if `render` is imported from test-utils file we won't need to pass store and wrap in router here

File: src/apps/sources/components/SourceForm.tsx [hunk 6]
```diff
@@ -225,28 +284,29 @@ const SourceForm: React.FC<Props> = ({
...
-                            // required
-                            fullWidth
-                            autoComplete="off"
-                            id="external_id"
-                            name="external_id"
-                            label="External Id"
-                            margin="normal"
-                            component={TextField}
-                        />
+                        {viewing ? (
```
Comment (rakesh-1010) [inline src/apps/sources/components/SourceForm.tsx]: can we extract this block in a function above, and call the function here. return will be much cleaner that way

File: src/apps/sources/pages/CreateSourcePage.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    usersOrgs?: APIOrg[];
+    createSourceAction: (
+        ...args: Parameters<typeof createSourceDispatchAction>
+    ) => void;
+    loading: boolean;
+    newSource?: APISource;
+}
+
+const CreateSourcePage: React.FC<Props> = ({
+                                                   profile,
```
Comment (rakesh-1010) [inline src/apps/sources/pages/CreateSourcePage.tsx]: fix indentation

File: src/apps/sources/pages/CreateSourcePage.tsx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    return (
+        <Grid id="create-source-page" item xs={6} component="div">
+            <Paper>
+                <SourceForm
+                    context={CONTEXT.create}
+                    errors={errors}
+                    profile={profile}
+                    usersOrgs={usersOrgs ? usersOrgs : []}
+                    loading={loading}
+                        onSubmit={(values: APISource) => createSourceAction(values)}
```
Comment (rakesh-1010) [inline src/apps/sources/pages/CreateSourcePage.tsx]: fix indentation

File: src/apps/sources/__test__/components/SourceForm.test.tsx [hunk 9]
```diff
@@ -2,40 +2,37 @@ import React from "react";
 import SourceForm from "../../components/SourceForm";
 import {APIOrg, APIProfile} from "../../../authentication";
 import {Source} from "../../types";
-import {render} from "@testing-library/react";
+import {render, fireEvent, act, wait, waitFor} from "@testing-library/react";
```
Comment (vinishad) [inline src/apps/sources/__test__/components/SourceForm.test.tsx]: @binduak @supriyatw can we remove wait and waitFor?

File: src/apps/sources/components/SourceForm.tsx [hunk 10]
```diff
@@ -16,6 +18,8 @@ import { snakeCase } from "lodash";
 
 import { Source } from "../types";
 import { APIOrg, APIProfile } from "../../authentication";
+import * as Yup from "yup";
+import {CONTEXT} from "../../dictionaries/constants";
```
Comment (vinishad) [inline src/apps/sources/components/SourceForm.tsx]: @binduak @supriyatw i think we are using wrong import of CONTEXT

File: src/apps/sources/__test__/pages/CreateSourcePage.test.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+import {APIOrg, APIProfile} from "../../../authentication";
+import {APISource, Source} from "../../types";
+import {createSourceAction, createSourceDispatchAction} from "../../redux";
+import {act, fireEvent, render} from "@testing-library/react";
+import {Provider} from "react-redux";
+import store from "../../../../redux";
+import {BrowserRouter as Router} from "react-router-dom";
+import {ViewSourcePage} from "../../pages";
+import {CONTEXT} from "../../constants";
+import {currentState, personalSources, testSource} from "../test_data";
```
Comment (vinishad) [inline src/apps/sources/__test__/pages/CreateSourcePage.test.tsx]: @binduak @supriyatw can we remove all the unused imports

File: src/apps/sources/__test__/components/SourceForm.test.tsx (hunk 9 above)
Comment (binduak) [reply src/apps/sources/__test__/components/SourceForm.test.tsx]: Done @vinishad

File: src/apps/sources/__test__/components/SourceForm.test.tsx (hunk 1 above)
Comment (binduak) [reply src/apps/sources/__test__/components/SourceForm.test.tsx]: Done @rakesh-1010

File: src/apps/sources/__test__/components/SourceForm.test.tsx (hunk 2 above)
Comment (binduak) [reply src/apps/sources/__test__/components/SourceForm.test.tsx:91]: @rakesh-1010 Done

File: src/apps/sources/__test__/pages/CreateSourcePage.test.tsx (hunk 3 above)
Comment (binduak) [reply src/apps/sources/__test__/pages/CreateSourcePage.test.tsx:63]: @rakesh-1010 Done !

File: src/apps/sources/__test__/pages/CreateSourcePage.test.tsx (hunk 4 above)
Comment (binduak) [reply src/apps/sources/__test__/pages/CreateSourcePage.test.tsx]: @rakesh-1010 Done !

File: src/apps/sources/__test__/pages/CreateSourcePage.test.tsx (hunk 5 above)
Comment (binduak) [reply src/apps/sources/__test__/pages/CreateSourcePage.test.tsx]: @rakesh-1010 Done !

File: src/apps/sources/components/SourceForm.tsx (hunk 6 above)
Comment (binduak) [reply src/apps/sources/components/SourceForm.tsx]: @rakesh-1010 Done !

File: src/apps/sources/pages/CreateSourcePage.tsx (hunk 8 above)
Comment (binduak) [reply src/apps/sources/pages/CreateSourcePage.tsx]: @rakesh-1010 Done !

File: src/apps/sources/components/SourceForm.tsx (hunk 10 above)
Comment (binduak) [reply src/apps/sources/components/SourceForm.tsx]: @rakesh-1010 Done !

File: src/apps/sources/__test__/pages/CreateSourcePage.test.tsx (hunk 11 above)
Comment (binduak) [reply src/apps/sources/__test__/pages/CreateSourcePage.test.tsx]: @rakesh-1010 Done !

**Review by rakesh-1010 — APPROVED** [review]

**Pull request merged**